### ISSUE TITLE
Processing ZIP files in Java broken with latest zlib

### ISSUE THREAD

**Issue by sithmein** [madler/zlib#305]
After the upgrade to macOS High Sierra several Java applications that compress data stopped working correctly. Example are [KNIME Analytics Platform](https://www.knime.com/forum/knime-general/high-sierra-and-node-out-zip-files) or [Tomcat](https://stackoverflow.com/questions/46539453/tomcat-with-compression-enabled-causes-error-on-os-x-high-sierra). I was able to track down the problem to a potential bug in the zlib version shipped with High Sierra (1.12.11). Version 1.2.5 which is used in Sierra works find. The bug is caused by changing the compression levels for different entries.
[CreateZip.java.txt](https://github.com/madler/zlib/files/1380603/CreateZip.java.txt)
 (rename to .java) demonstrates the problem. Compile and start with _/tmp 100_ as arguments.
It works fine on any operating system except macOS High Sierra. If you comment the three lines that are marked with "Triggers the bug" the program also works as expected.
This is a blocker for any Java application that changes compression levels on macOS therefore please consider this a critical bug.

**Comment by xuemingshen** [comment]
see #275

**Comment by madler** [comment]
Thanks. I'll take a look.

**Comment by madler** [comment]
This was fixed in https://github.com/madler/zlib/commit/f9694097dd69354b03cb8af959094c7f260db0a1 .
Can you verify?

**Comment by sithmein** [comment]
I still get the same problem with the latest code from the master or develop branch.

**Comment by madler** [comment]
What happens?

**Comment by sithmein** [comment]
```
knime400:~ jenkins$ java -classpath . CreateZip /tmp/ 100
Writing 100MB to /tmp/knime_container_2017-10-13_1864572657.bin... done (19s)
Reading it back in ... java.util.zip.ZipException: invalid block type
        at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:164)
        at java.util.zip.ZipInputStream.read(ZipInputStream.java:194)
        at java.io.FilterInputStream.read(FilterInputStream.java:107)
        at CreateZip.read(CreateZip.java:78)
        at CreateZip.main(CreateZip.java:122)
```

**Comment by madler** [comment]
I'll need the sequence of zlib calls made by the test program.

**Comment by sithmein** [comment]
[zlib.log](https://github.com/madler/zlib/files/1383220/zlib.log)
This wasn't easy since the functions are called from Java. I added some manual debug output to deflate.c, the result is attached. Let me know if you need more.

**Comment by xuemingshen** [comment]
from java's perspective, the test case should call s.closeEntry() before calling s.etLevel(...) + s.putNextEtnry() for the next entry. s.putNextEntry() calls s.closeEntry() if it is not invoked for the "last entry". In this case, it triggers the zlib to compress any leftover with  a different "level".

**Comment by sithmein** [comment]
Good catch, that seems to resolve the problem. Still the question remains why it worked without explicitly closing. Attached the log for the same run
[zlib2.log](https://github.com/madler/zlib/files/1383254/zlib2.log)
 but with explicitly closing entries before opening new ones.

**Comment by madler** [comment]
@sithmein Thank you very much for your effort! Could you also provide the values of `avail_in` and `avail_out` before and after each `deflate()` and `deflateParams()` call?

**Comment by sithmein** [comment]
Sure!
[zlib-with-avail-no-closeEntry.txt](https://github.com/madler/zlib/files/1383311/zlib-with-avail-no-closeEntry.txt)
[zlib-with-avail-with-closeEntry.txt](https://github.com/madler/zlib/files/1383312/zlib-with-avail-with-closeEntry.txt)

**Comment by xuemingshen** [comment]
@madler in Java, if the "level" is changed in the middle of compression, we call "deflateParam" with the new level/stragegy to reset them. In this case,  deflateParam returns Z_BUF_ERROR in 1.2.11, even it did flush the leftover bytes (using the previous level) into output buffer. Java wrapper code however has been interpreting the "Z_BUF_ERROR" as an "error" and does not expect anything actually getting flushed into the output buf (so it does check and update its output buffer off/len ...).  The previous release does not appear to return Z_BUF_ERROR. It returns the err code from deflate(...) directly and it appears to be Z_OK in this case.  So I would assume this is an incompatible change (?). Based on the spec, it might appear to be OK to return either Z_OK or Z_BUF_ERROR, though.

**Comment by xuemingshen** [comment]
(so it does check and update its output buffer off/len ...). ---> so it does NOT check and ...

**Comment by madler** [comment]
Thank you for the `avail_in` values. They don't look like they make sense though. They would not be zero going into all of those `deflate()` calls. Can you check when you are fetching those?

What I would like is `avail_in` as well as `avail_out` immediately before each `deflate()` and `deflateParams()` call, and those values immediately after each call.

**Comment by madler** [comment]
@xuemingshen If `deflateParams()` returns `Z_BUF_ERROR`, then there was not enough output space to compress all of the provided input before applying the parameter change. In that case, the parameters _are not changed_. You are expected to call `deflateParams()` again with more output space and the same requested parameters until `Z_BUF_ERROR` is not returned. When that happens, the parameters are changed.

This can all be avoided by instead processing all of the available input using `deflate()` calls with sufficient output space. Then the `deflateParams()` call with no available input will simply succeed.

**Comment by madler** [comment]
@xuemingshen Can you point me to the code in the JDK for the `setLevel()` method?

**Comment by sithmein** [comment]
Actually I think, the number for `avail_in` do make sense. The Java code writes three times two blocks of 16kB (first with compression, second without, third with). I don't know where all the other calls to `deflate()` originate from. Some may result from flushing the stream on the Java side and closing entries. However writing 0 bytes there seems reasonable as there are no more bytes to write.

http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/33c7189b9693/src/share/native/java/util/zip/Deflater.c is the source of the JNI bridge code to zlib. setLevel in Java (see http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/util/zip/Deflater.java/) doesn't result in an immediate call to zlib but only sets a `level` value which results in a call to `deflateParams()` upon writing the next bunch of data.

**Comment by sithmein** [comment]
BTW, the commit that broke it was [7161ad7](https://github.com/madler/zlib/commit/7161ad76e2d0ac7de2a6235fcad3b9dfc99e9140)

**Comment by xuemingshen** [comment]
@madler 

[the diff, left side is the original/existing code, right is the version I will put in to fix the problem]
http://cr.openjdk.java.net/~sherman/zlib1.2.11.fix/src/java.base/share/native/libzip/Deflater.c.sdiff.html

[original/existing version]
http://cr.openjdk.java.net/~sherman/zlib1.2.11.fix/src/java.base/share/native/libzip/Deflater.c-.html

[new version that fixed the issue]
http://cr.openjdk.java.net/~sherman/zlib1.2.11.fix/src/java.base/share/native/libzip/Deflater.c.html

Java wrapper does misinterpret the return value "Z_BUF_ERROR" from very beginning (probably since its first implementation), probably back decade ago, based on the (old) api doc

--------------------
     deflateParams returns Z_OK if success, Z_STREAM_ERROR if the source                                                                     
   stream state was inconsistent or if a parameter was invalid, Z_BUF_ERROR if                                                               
   strm->avail_out was zero.    
--------------------    

which was interpreted as "returns Z_BUF_ERROR if the input parameter avail_out was zero, therefor nothing can be written out". So the implementation does not try to check if there is any new bytes written into the output buffer (when Z_BUF_ERROR returned) and update the corresponding output buffer's offset and len, if there is any new output bytes (the direct consequence is that this part of the output bytes were be overwritten by coming new bytes), as the implementation does when the return value is Z_OK. This assumption appears to be WRONG. But the previous zlib implementation did not return Z_BUF_ERROR in its previous releases (it simply returns the "err" from the deflate(...) that flushes the internal buffer, with old "level" in deflateParams()). So we never catches this issue. That said the logic appears to be incorrect in the old/existing implementation to update the "setParamsID" back to false in this case (which means the "level" actually is not getting updated in underlying deflater). Again, since we did not hit this code/path in early release, it has not been noticed till now. And I do have a bug report against this behavior.

It appears to be the behavior change in the 1.2.11 (to be clear I am NOT saying the behavior is wrong) hits a "old" mis-implementation/interpretation in Java zlib wrapper code. I will try to put the fix/change out asap, including the update releases for the previous releases, if possible. But unfortunately it will take time and the end Java user will start to hit this issue when 1.2.11 is being popped up as the default zlib for the upcoming new release, including various linux distributions.

Here is the info regarding what happened. Those deflate invocations after deflateParams appears to be flush the bytes out of the internal output buffer. The default output buffer for java is 512.
-------------------------------------------------------------

Writing 100MB to /tmp/knime_container_2017-10-13.zip...

... [normall compression with > 1 level]

deflateParams: level=0, strategy=0  next_in=ecdefd80, avail_in=0, avail_out=512        --> ret=-5  z_buf_error
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     ok
deflate: next_in=ecdefd80, avail_in=0, avail_out=512, flush=4 -->     z_stream_end

**Comment by madler** [comment]
@xuemingshen Thank you very much for your work on this. Indeed it appears that Java will inadvertently discard compressed data, which is what results in the corrupted output seen in the reports. It would be great if you could post a comment back here when the JDK has been repaired.

Yes, I see now how there could be so many calls to `deflate()` with `avail_in` zero, since the output buffer is so incredibly small! 512 bytes. Applications using zlib would be faster if the buffer size were larger. This is especially true for `inflate()`, which uses special high-speed code for large input and output buffers. Is that 512 something specified inside the JDK, or did it come from the application?

**Comment by madler** [comment]
As a fix for the JDK has been identified, I am closing this issue. (Comments can still be made.)

**Comment by sithmein** [comment]
Thanks guys, I really appreciate the productive and quick handling of this issue.

**Comment by madler** [comment]
@xuemingshen I wonder about how the Java SDK implements `setLevel`. A quick perusal of the code indicates that the `deflateParams` call is deferred until the next time data is compressed. That's perfectly fine, but I wonder if `deflateParams` is then being fed new data that was presented after the `setLevel` call? If so, the parameter change will be delayed until such time as there is enough output space to write all of the compressed output for that particular call. If you call `deflateParams` again with yet more data, it will continue to wait until it has all been compressed and written. With small output buffers and continuing to provide more input data, that could be a long time, which is not what the user would have intended.

What should be done (and I can't tell exactly what's happening with my quick perusal), is that when a `setLevel` is requested, all of the input data presented up to that point, but no further, should be compressed and consumed. That can be done either with `deflate` or `deflateParams` until `avail_out` is not zero. Then a `deflateParams` will have immediate effect. Only after that should `deflate` be called with new input data that was presented after the `setLevel` call.

So do you know how the JDK behaves in this regard?

**Comment by xuemingshen** [comment]
@madler 

We do have an  issue here on how to handle the "level change in the middle of compression" in Java now.

Until now (before 1.2.11) zlib's deflateParams() is implemented as showed below. So if it is called with a different level, the zlib internal "level" in s->level is being updated to the new value, regardless of the returned result from "flush the last buffer" / err = deflate(stem, Z_BLOCK), Z_OK or Z_BUF_ERROR.

deflateParams(...) {
...
    if (level < 0 || level > 9 || strategy < 0 || strategy > Z_FIXED) {
        return Z_STREAM_ERROR;
    }
    func = configuration_table[s->level].func;
    if ((strategy != s->strategy || func != configuration_table[level].func) &&
        strm->total_in != 0) {
        /* Flush the last buffer: */
        err = deflate(strm, Z_BLOCK);
    }
    if (s->level != level) {
        s->level = level;
        s->max_lazy_match   = configuration_table[level].max_lazy;
        s->good_match       = configuration_table[level].good_length;
        s->nice_match       = configuration_table[level].nice_length;
        s->max_chain_length = configuration_table[level].max_chain;
    }
    s->strategy = strategy;
    return err;
}

Given this,  the old/existing java wrapper code calls/invokes deflateParams(...) only once overtime the "level" is changed via Java API Deflater.setLevel() (not immediately, but when next time deflate() is invoked, as you noticed, but it does set a flag "setParams" to true). So Deflater.java.deflate() -> Deflate.c/deflateBytes()  calls deflateParams() when the "setParams" is on, AND turns this flag off when the deflateParams() return Z_OK/Z_BUF_ERR. So next Deflater.deflate(...) will go to deflate(...) not deflateParams() any more.

This might result in the "unexpected" behavior, as you suggested,that if there is still leftover bytes in Java Deflater's input buffer, those bytes might/will be compressed with the new level, , if zlib's internal buffer is full. But this might not treated as a fatal error, from runtime point of view, given the fact the bytes are not corrupted. As you suggested the only way to make it actually work (with the old zlib implementation) is to keep calling deflate() with the leftover bytes + old "level" till all bytes are out. But what does it mean is actually we have to pull the logic implemented in deflateParams() to Java level and implement it ourself. 

That said, the Java de/inflater APIs have been like this for 20 years :-)  and this has NOT been a problem in real world , as normally we have a use pattern like

def.setInput(in_buf, in_buf_off, in_buf_len);
while (in_len > 0) {       (in_len is being updated from native Deflater.c/deflateBytes)
    int len = def.deflate(out_buff, 0, out_buf_len);
    if (len > 0)
       out.write(out_buf, 0, len);
}

And it appears in most user scenarios level/strategy the level/strategy is always being set after all existing "input bytes" are compressed (for example, in above case in_len>=0). The Java Deflater use extra flags "finish", "finished" and "needsInput" to help controlling the deflating sequence. This might not be the best APIs, but given Java's strong "compatibility" request, it's hard to touch it without risking breaking someone else's running code. I need take a deep look to see if it's possible to implement this without breaking anything.

**Comment by sithmein** [comment]
Turns out, simply calling `closeEntry` before changing the compression level doesn't solve the problem in all cases. It works for the simple program I attached above but in our real application we are still seeing the issue. @xuemingshen , given your expertise in the native code binding is there any workaround on the pure Java we could apply in order to work around the problem (not changing the compression level is not an option)? Are there any estimates when the bug in the JDK will be fixed? Is there a bug entry for OpenJDK?

**Comment by xuemingshen** [comment]
@sithmein I need a specific test case to comment if there is any workaround from Java site. We are facing two issues here. (1) the issue being fixed/addressed in  zlib #f969409 (2) the java wrapper code needs to deal with the "incompatible" change since zlib1.2.9-1.2.11. (deflateParams() returns z_buf_error). (1) has been fix in zlib repo but it's not out yet and I'm still working on (2). If you can use a private zlib binary in your env, the alternative might be to downgrade to zlib1.2.5, the zlib binary used in previous macos ?

**Comment by xuemingshen** [comment]
@madler

I took a quick looks at all releases and were trying to have a "unified" workaround. Seems like it's "difficult" to workaround for the releases 1.2.3 and earlier, which does not appears to work well with the "strategy-only-change-in-the-middle-of compression", even tried to use "partial-flush" to replace the z_block. I know these are decade-old releases but they are still showing up on various linux distributions :-(

1.2.3:

int ZEXPORT deflateParams(strm, level, strategy)

    ...
    if (func != configuration_table[level].func && strm->total_in != 0) {
        /* Flush the last buffer: */
        err = deflate(strm, Z_PARTIAL_FLUSH);
    }
    if (s->level != level) {
    ...
}

1.2.4
int ZEXPORT deflateParams(strm, level, strategy)

    ...

    if ((strategy != s->strategy || func != configuration_table[level].func) &&
        strm->total_in != 0) {
        /* Flush the last buffer: */
        err = deflate(strm, Z_BLOCK);
    }
    if (s->level != level) {
    ...

1.2.5:

int ZEXPORT deflateParams(strm, level, strategy) {

    ...

    if ((strategy != s->strategy || func != configuration_table[level].func) &&
        strm->total_in != 0) {
        /* Flush the last buffer: */
        err = deflate(strm, Z_BLOCK);
    }
    if (s->level != level) {
    ...
}

1.2.8:

int ZEXPORT deflateParams(strm, level, strategy) {

    ...

    if ((strategy != s->strategy || func != configuration_table[level].func) &&
        strm->total_in != 0) {
        /* Flush the last buffer: */
        err = deflate(strm, Z_BLOCK);
        if (err == Z_BUF_ERROR && s->pending == 0)
            err = Z_OK;
    }
    if (s->level != level) {
    ...
}

1.2.9/10:

int ZEXPORT deflateParams(strm, level, strategy) {

    ...
    if ((strategy != s->strategy || func != configuration_table[level].func)) {
        /* Flush the last buffer: */
        int err = deflate(strm, Z_BLOCK);
        if (err == Z_STREAM_ERROR)
            return err;
        if (strm->avail_out == 0)
            return Z_BUF_ERROR;
    }
    if (s->level != level) {
    ...
}

1.2.11

int ZEXPORT deflateParams(strm, level, strategy) {
    ....

    if ((strategy != s->strategy || func != configuration_table[level].func) &&
        s->high_water) {
        /* Flush the last buffer: */
        int err = deflate(strm, Z_BLOCK);
        if (err == Z_STREAM_ERROR)
            return err;
        if (strm->avail_out == 0)
            return Z_BUF_ERROR;
    }
    if (s->level != level) {
    ...
}

1.2.12  (? latest repo)

int ZEXPORT deflateParams(strm, level, strategy) {
    ....

    if ((strategy != s->strategy || func != configuration_table[level].func) &&
        s->last_flush != -2) {
        /* Flush the last buffer: */
        int err = deflate(strm, Z_BLOCK);
        if (err == Z_STREAM_ERROR)
            return err;
        if (strm->avail_out == 0)
            return Z_BUF_ERROR;
    }
    if (s->level != level) {
    ...
}

**Comment by sithmein** [comment]
@xuemingshen , for a testcase see this snippet:

```
private static boolean isZLIBSupportsLevelSwitch() {
    ByteArrayOutputStream byteArrayOut;
    byte[] nullBytes = new byte[1024 * 1024];
    try (ZipOutputStream zipOut = new ZipOutputStream(byteArrayOut = new ByteArrayOutputStream())) {
        zipOut.setLevel(Deflater.BEST_SPEED);
        zipOut.putNextEntry(new ZipEntry("deflated.bin"));
        zipOut.write(nullBytes);
        zipOut.closeEntry();
        zipOut.putNextEntry(new ZipEntry("stored.bin"));
        zipOut.setLevel(Deflater.BEST_COMPRESSION);
        zipOut.write(nullBytes);
        zipOut.closeEntry();
    } catch (IOException e) {
        LOGGER.error("Unexpected error creating test zipped output", e);
        return false;
    }
    try (ZipInputStream zipIn = new ZipInputStream(new ByteArrayInputStream(byteArrayOut.toByteArray()))) {
        for (int i = 0; i < 2; i++) {
            zipIn.getNextEntry();
            while (zipIn.read(nullBytes) >= 0) {
            }
        }
    } catch (ZipException e) {
        return false;
    } catch (IOException e) {
        LOGGER.error("Unexpected error creating test zipped output", e);
        return false;
    }
    return true;
}
```

**Comment by LanceAndersen** [comment]
Can you point me to where this fix is integrated as I did not see it in the current codeline
